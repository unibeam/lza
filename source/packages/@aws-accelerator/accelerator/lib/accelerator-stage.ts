/**
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance
 *  with the License. A copy of the License is located at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions
 *  and limitations under the License.
 */
export enum AcceleratorStage {
  PIPELINE = 'pipeline',
  /**
   * Accelerator Tester Pipeline
   */
  TESTER_PIPELINE = 'tester-pipeline',
  /**
   * Prepare Stage - Verify the configuration files, environment and create accounts
   */
  PREPARE = 'prepare',
  /**
   * DiagnosticsPack Stage - Creates Diagnostics pack resources
   */
  DIAGNOSTICS_PACK = 'diagnostics-pack',
  ORGANIZATIONS = 'organizations',
  KEY = 'key',
  LOGGING = 'logging',
  /**
   * Accounts Stage - Handle all Organization and Accounts actions
   */
  ACCOUNTS = 'accounts',
  BOOTSTRAP = 'bootstrap',
  CUSTOMIZATIONS = 'customizations',
  DEPENDENCIES = 'dependencies',
  SECURITY = 'security',
  SECURITY_RESOURCES = 'security-resources',
  OPERATIONS = 'operations',
  IDENTITY_CENTER = 'identity-center',
  NETWORK_PREP = 'network-prep',
  NETWORK_VPC = 'network-vpc',
  NETWORK_VPC_ENDPOINTS = 'network-vpc-endpoints',
  NETWORK_VPC_DNS = 'network-vpc-dns',
  NETWORK_ASSOCIATIONS = 'network-associations',
  NETWORK_ASSOCIATIONS_GWLB = 'network-associations-gwlb',
  SECURITY_AUDIT = 'security-audit',
  RESOURCE_POLICY_ENFORCEMENT = 'resource-policy-enforcement',
  FINALIZE = 'finalize',
  IMPORT_ASEA_RESOURCES = 'import-asea-resources',
  POST_IMPORT_ASEA_RESOURCES = 'post-import-asea-resources',
}
