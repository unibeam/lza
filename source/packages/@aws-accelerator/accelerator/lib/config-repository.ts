/**
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance
 *  with the License. A copy of the License is located at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions
 *  and limitations under the License.
 */

import {
  AccountsConfig,
  GlobalConfig,
  IamConfig,
  NetworkConfig,
  OrganizationConfig,
  SecurityConfig,
  Region,
  ControlTowerLandingZoneConfig,
} from '@aws-accelerator/config';
import { Bucket, BucketEncryptionType } from '@aws-accelerator/constructs';
import { NagSuppressions } from 'cdk-nag';
import * as cdk_extensions from '@aws-cdk-extensions/cdk-extensions';
import * as cdk from 'aws-cdk-lib';
import * as s3_assets from 'aws-cdk-lib/aws-s3-assets';
import { Construct } from 'constructs';
import * as fs from 'fs';
import * as yaml from 'js-yaml';
import * as os from 'os';
import * as path from 'path';
import AdmZip from 'adm-zip';

export interface ConfigRepositoryProps {
  readonly description?: string;
  readonly managementAccountEmail: string;
  readonly logArchiveAccountEmail: string;
  readonly auditAccountEmail: string;
  readonly controlTowerEnabled: string;
  readonly enableSingleAccountMode: boolean;
  /**
   * AWS Control Tower Landing Zone configuration
   */
  readonly controlTowerLandingZoneConfig?: ControlTowerLandingZoneConfig;
}

export interface CodeCommitConfigRepositoryProps extends ConfigRepositoryProps {
  readonly repositoryName: string;
  readonly repositoryBranchName?: string;
}
export interface S3ConfigRepositoryProps extends ConfigRepositoryProps {
  readonly configBucketName: string;
  readonly installerKey: cdk.aws_kms.Key;
  readonly serverAccessLogsBucketName: string;
  readonly autoDeleteObjects?: boolean;
}

/**
 * Class to create AWS accelerator configuration repository and initialize the repository with default configuration
 */
export class ConfigRepository extends Construct {
  readonly tempDirPath: string;
  constructor(scope: Construct, id: string, props: ConfigRepositoryProps) {
    super(scope, id);

    //
    // Generate default configuration files
    //
    this.tempDirPath = fs.mkdtempSync(path.join(os.tmpdir(), 'config-assets-'));

    let controlTowerEnabledValue = true;
    let managementAccountAccessRole = 'AWSControlTowerExecution';
    if (props.controlTowerEnabled.toLowerCase() === 'no') {
      controlTowerEnabledValue = false;
      managementAccountAccessRole = 'OrganizationAccountAccessRole';
    }

    fs.writeFileSync(
      path.join(this.tempDirPath, GlobalConfig.FILENAME),
      yaml.dump(
        new GlobalConfig({
          homeRegion: cdk.Stack.of(this).region as Region,
          controlTower: { enable: controlTowerEnabledValue, landingZone: props.controlTowerLandingZoneConfig },
          managementAccountAccessRole: managementAccountAccessRole,
        }),
      ),
      'utf8',
    );

    fs.writeFileSync(
      path.join(this.tempDirPath, AccountsConfig.FILENAME),
      yaml.dump(
        new AccountsConfig({
          managementAccountEmail: props.managementAccountEmail,
          logArchiveAccountEmail: props.logArchiveAccountEmail,
          auditAccountEmail: props.auditAccountEmail,
        }),
      ),
      'utf8',
    );

    fs.writeFileSync(path.join(this.tempDirPath, IamConfig.FILENAME), yaml.dump(new IamConfig()), 'utf8');
    fs.writeFileSync(path.join(this.tempDirPath, NetworkConfig.FILENAME), yaml.dump(new NetworkConfig()), 'utf8');
    if (props.enableSingleAccountMode) {
      const orgConfig = new OrganizationConfig({
        enable: false,
        organizationalUnits: [
          {
            name: 'Security',
            ignore: undefined,
          },
          {
            name: 'LogArchive',
            ignore: undefined,
          },
        ],
        organizationalUnitIds: [],
        serviceControlPolicies: [],
        taggingPolicies: [],
        backupPolicies: [],
      });
      fs.writeFileSync(path.join(this.tempDirPath, OrganizationConfig.FILENAME), yaml.dump(orgConfig), 'utf8');
    } else {
      fs.writeFileSync(
        path.join(this.tempDirPath, OrganizationConfig.FILENAME),
        yaml.dump(new OrganizationConfig()),
        'utf8',
      );
    }

    fs.writeFileSync(path.join(this.tempDirPath, SecurityConfig.FILENAME), yaml.dump(new SecurityConfig()), 'utf8');
  }
}

export class CodeCommitConfigRepository extends ConfigRepository {
  readonly configRepo: cdk_extensions.Repository;
  readonly s3AssetBucket: cdk.aws_s3_assets.Asset;
  constructor(scope: Construct, id: string, props: CodeCommitConfigRepositoryProps) {
    super(scope, id, props);

    this.s3AssetBucket = new s3_assets.Asset(this, 'ConfigurationDefaultsAssets', {
      path: this.tempDirPath,
    });
    this.configRepo = new cdk_extensions.Repository(this, 'Resource', {
      repositoryName: props.repositoryName,
      repositoryBranchName: props.repositoryBranchName!,
      s3BucketName: this.s3AssetBucket.bucket.bucketName,
      s3key: this.s3AssetBucket.s3ObjectKey,
    });
  }
  /**
   * Method to get initialized CodeCommit repository object
   * @return Returns Initialized CodeCommit repository object.
   */
  public getRepository(): cdk_extensions.Repository {
    return this.configRepo;
  }
}

export class S3ConfigRepository extends ConfigRepository {
  readonly configRepo: Bucket;
  constructor(scope: Construct, id: string, props: S3ConfigRepositoryProps) {
    super(scope, id, props);

    this.configRepo = new Bucket(this, 'SecureBucket', {
      encryptionType: BucketEncryptionType.SSE_KMS,
      s3BucketName: props.configBucketName,
      kmsKey: props.installerKey,
      serverAccessLogsBucketName: props.serverAccessLogsBucketName,
      autoDeleteObjects: props.autoDeleteObjects,
      s3LifeCycleRules: [],
    });

    this.getZippedConfigFiles();

    /**
     * The default LZA configuration must first be uploaded to a separate path in the bucket.
     * This lowers the risk of overwriting customer configuration zip files when the default LZA config files change
     *
     * https://docs.aws.amazon.com/cdk/api/v2/docs/aws-cdk-lib.aws_s3_deployment-readme.html#notes
     *   If you are using s3deploy.Source.bucket() to take the file source from another bucket: the deployed files will only be updated if the key (file name) of the file in the source bucket changes. Mutating the file in place will not be good enough: the custom resource will simply not run if the properties don't change.
     *   If you use assets (s3deploy.Source.asset()) you don't need to worry about this: the asset system will make sure that if the files have changed, the file name is unique and the deployment will run.
     */
    const defaultDeployment = new cdk.aws_s3_deployment.BucketDeployment(this, 'UploadDefaultZipFileToS3', {
      sources: [cdk.aws_s3_deployment.Source.asset(`${this.tempDirPath}/zipped`)],
      destinationBucket: this.configRepo.getS3Bucket(),
      destinationKeyPrefix: 'default',
      prune: false,
    });

    /**
     * CAUTION: This BucketDeployment controls the creation of LZA config files in S3.
     * Modification of this resource has the potential to overwrite existing customer LZA configurations.
     * Please see the following page for more information:
     * https://docs.aws.amazon.com/cdk/api/v2/docs/aws-cdk-lib.aws_s3_deployment.BucketDeployment.html
     *
     * Copy default config to /zipped prefix to use as CodePipeline source
     * */
    const s3CopyDeployment = new cdk.aws_s3_deployment.BucketDeployment(this, 'CopyDefaultZipToTarget', {
      sources: [
        cdk.aws_s3_deployment.Source.bucket(this.configRepo.getS3Bucket(), 'default/aws-accelerator-config.zip'),
      ],
      destinationBucket: this.configRepo.getS3Bucket(),
      destinationKeyPrefix: 'zipped',
      extract: false,
      prune: false,
    });
    s3CopyDeployment.node.addDependency(defaultDeployment);

    const cdkBucketDeploymentIds = [];
    for (const child of cdk.Stack.of(this).node.children) {
      if (child.node.id.startsWith('Custom::CDKBucketDeployment')) {
        cdkBucketDeploymentIds.push(child.node.id);
      }
    }

    for (const cdkBucketDeploymentId of cdkBucketDeploymentIds) {
      NagSuppressions.addResourceSuppressionsByPath(
        cdk.Stack.of(this),
        `${cdk.Stack.of(this).stackName}/${cdkBucketDeploymentId}/ServiceRole/Resource`,
        [
          {
            id: 'AwsSolutions-IAM4',
            reason: 'CDK construct auto-generated role.',
          },
        ],
      );
    }
  }

  /**
   * Method to create a zip file of LZA config files
   *
   * @return Returns the path to the zip file
   */
  public getZippedConfigFiles(): string {
    const configZipFilePath = `${this.tempDirPath}/zipped/aws-accelerator-config.zip`;
    const admZip = new AdmZip();

    const files = fs.readdirSync(this.tempDirPath);
    // Add files to the zip archive
    for (const file of files) {
      const filePath = path.join(this.tempDirPath, file);
      const fileData = fs.readFileSync(filePath);
      const fileName = path.basename(filePath);
      admZip.addFile(fileName, fileData);
    }

    // Write the zip file to disk
    admZip.writeZip(configZipFilePath);
    return configZipFilePath;
  }

  /**
   * Method to get initialized S3 repository object
   *
   * @return Returns Initialized S3 repository object.
   */
  public getRepository(): cdk.aws_s3.IBucket {
    return this.configRepo.getS3Bucket();
  }
}
