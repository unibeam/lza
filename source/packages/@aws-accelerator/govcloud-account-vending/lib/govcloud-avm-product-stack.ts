/**
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance
 *  with the License. A copy of the License is located at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions
 *  and limitations under the License.
 */

import * as cdk from 'aws-cdk-lib';
import { Construct } from 'constructs';

export interface GovCloudAccountVendingProductStackProps extends cdk.StackProps {
  readonly acceleratorPrefix: string;
}

export class GovCloudAccountVendingProductStack extends cdk.aws_servicecatalog.ProductStack {
  constructor(scope: Construct, id: string, props: GovCloudAccountVendingProductStackProps) {
    super(scope, id);

    /** This stack creates service catalog product which can 
     * create AWS GovCloud (US) accounts using the Organizations API
    * "CreateGovCloudAccount". 
    * 
    * 
    * The account where this stack is launched should 
    * 1. be in commercial region with Organizations enabled
    * 2. Have ability to create AWS GovCloud (US) Accounts
    * Please read https://docs.aws.amazon.com/govcloud-us/latest/UserGuide/getting-set-up.html
    * and the API documentation 
    * https://docs.aws.amazon.com/organizations/latest/APIReference/API_CreateGovCloudAccount.html

    */

    // Parameters for account creation
    const accountName = new cdk.CfnParameter(this, 'AccountName', {
      type: 'String',
      description:
        "The friendly name of the member account.  The account name can consist of only the characters [a-z],[A-Z],[0-9], hyphen (-), or dot (.) You can't separate characters with a dash (â€“).",
    });
    const emailAddress = new cdk.CfnParameter(this, 'AccountEmail', {
      type: 'String',
      description:
        'Specifies the email address of the owner to assign to the new member account in the commercial Region. This email address must not already be associated with another AWS account. You must use a valid email address to complete account creation.',
    });
    const orgAccessRole = new cdk.CfnParameter(this, 'OrgAccessRole', {
      type: 'String',
      description:
        'The name of an IAM role that AWS Organizations automatically preconfigures in the new member accounts in both the AWS GovCloud (US) Region and in the commercial Region. This role trusts the management account, allowing users in the management account to assume the role, as permitted by the management account administrator. The role has administrator permissions in the new member account.',
      default: 'OrganizationAccountAccessRole',
    });
    const parameterGroups: { Label: { default: string }; Parameters: string[] }[] = [
      {
        Label: { default: 'Account Configuration' },
        Parameters: [accountName.logicalId, emailAddress.logicalId, orgAccessRole.logicalId],
      },
    ];
    const GovCloudAccountVendingParameterLabels: { [p: string]: { default: string } } = {
      [accountName.logicalId]: { default: 'Account Name' },
      [emailAddress.logicalId]: { default: 'Account Email' },
      [orgAccessRole.logicalId]: { default: 'Organization Role Name' },
    };
    // Parameter Metadata
    this.templateOptions.metadata = {
      'AWS::CloudFormation::Interface': {
        ParameterGroups: parameterGroups,
        ParameterLabels: { ...GovCloudAccountVendingParameterLabels },
      },
    };

    const accountVendingFunction = cdk.aws_lambda.Function.fromFunctionName(
      this,
      'AccountVendingFunction',
      `${props.acceleratorPrefix}-GovCloudAccountVending`,
    );

    //
    // Custom Resource definition
    //
    const resource = new cdk.CustomResource(this, 'CustomCreateGovCloudAccount', {
      resourceType: 'Custom::CreateGovCloudAccount',
      serviceToken: accountVendingFunction.functionArn,
      properties: {
        accountName: accountName.valueAsString,
        emailAddress: emailAddress.valueAsString,
        orgAccessRole: orgAccessRole.valueAsString,
      },
    });

    new cdk.CfnOutput(this, 'GovCloudAccountId', { value: resource.getAtt('GovCloudAccountId').toString() });
    new cdk.CfnOutput(this, 'AccountId', { value: resource.getAtt('AccountId').toString() });
  }
}
