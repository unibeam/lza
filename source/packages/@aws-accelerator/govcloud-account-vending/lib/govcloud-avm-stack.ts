/**
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance
 *  with the License. A copy of the License is located at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions
 *  and limitations under the License.
 */

import * as cdk from 'aws-cdk-lib';
import { Construct } from 'constructs';
import * as path from 'path';
import { GovCloudAccountVendingProductStack } from './govcloud-avm-product-stack';
import * as fs from 'fs';
import { version } from '../../../../package.json';

export interface GovCloudAccountVendingStackProps extends cdk.StackProps {
  readonly acceleratorPrefix: string;
}

export class GovCloudAccountVendingStack extends cdk.Stack {
  constructor(scope: Construct, id: string, props: GovCloudAccountVendingStackProps) {
    super(scope, id, props);

    /** This stack creates service catalog product which can 
     * create AWS GovCloud (US) accounts using the Organizations API
    * "CreateGovCloudAccount". 
    * 
    * 
    * The account where this stack is launched should 
    * 1. be in commercial region with Organizations enabled
    * 2. Have ability to create AWS GovCloud (US) Accounts
    * Please read https://docs.aws.amazon.com/govcloud-us/latest/UserGuide/getting-set-up.html
    * and the API documentation 
    * https://docs.aws.amazon.com/organizations/latest/APIReference/API_CreateGovCloudAccount.html

    */

    /*
     * Parameter for getting IAM Role ARN
     * This role will have access to launch service catalog products
     */

    // Create a portfolio
    const portfolio = new cdk.aws_servicecatalog.Portfolio(this, 'GovCloudAccountVendingPortfolio', {
      displayName: 'Landing Zone Accelerator on AWS',
      providerName: 'AWS Solutions',
    });

    // Create a GovCloud Account Vending service catalog product
    const product = new cdk.aws_servicecatalog.CloudFormationProduct(this, 'GovCloudAccountVendingProduct', {
      productName: 'Landing Zone Accelerator on AWS - AWS GovCloud (US) Account Vending',
      owner: 'AWS Solutions',
      productVersions: [
        {
          cloudFormationTemplate: cdk.aws_servicecatalog.CloudFormationTemplate.fromProductStack(
            new GovCloudAccountVendingProductStack(this, 'GovCloudAccountVendingProductStack', {
              acceleratorPrefix: props.acceleratorPrefix,
              description: `(SO0199-govcloudavmproduct) Landing Zone Accelerator on AWS. Version ${version}.`,
            }),
          ),
          productVersionName: 'v1.0.0',
          description:
            'AWS GovCloud (US) Account Vending Product. Create AWS GovCloud (US) accounts. Required inputs are Account name, email and Organization Access Role.',
        },
      ],
    });
    // Associate product to the portfolio
    portfolio.addProduct(product);

    const fileContents = fs.readFileSync(path.join(__dirname, 'lambdas/create-govcloud-account/index.js'));

    // Lambda function to be used in Custom Resource
    const accountVendingFunction = new cdk.aws_lambda.Function(this, 'GovCloudAccountVendingFunction', {
      code: new cdk.aws_lambda.InlineCode(fileContents.toString()),
      runtime: cdk.aws_lambda.Runtime.NODEJS_18_X,
      handler: 'index.handler',
      timeout: cdk.Duration.seconds(900),
      functionName: `${props.acceleratorPrefix}-GovCloudAccountVending`,
      description: 'Create AWS GovCloud (US) Accounts',
      initialPolicy: [
        new cdk.aws_iam.PolicyStatement({
          actions: ['organizations:CreateGovCloudAccount', 'organizations:DescribeCreateAccountStatus'],
          resources: ['*'],
        }),
      ],
    });

    //
    // cfn-nag suppressions
    //
    const cfnLambdaFunctionPolicy = accountVendingFunction.node.findChild('ServiceRole').node.findChild('DefaultPolicy')
      .node.defaultChild as cdk.aws_iam.CfnPolicy;
    cfnLambdaFunctionPolicy.cfnOptions.metadata = {
      cfn_nag: {
        rules_to_suppress: [
          {
            id: 'W12',
            reason: `IAM policy should not allow * resource.`,
          },
        ],
      },
    };

    const cfnLambdaFunction = accountVendingFunction.node.defaultChild as cdk.aws_lambda.CfnFunction;
    cfnLambdaFunction.cfnOptions.metadata = {
      cfn_nag: {
        rules_to_suppress: [
          {
            id: 'W58',
            reason: `CloudWatch Logs are enabled in AWSLambdaBasicExecutionRole`,
          },
          {
            id: 'W89',
            reason: `This function supports infrastructure deployment and is not deployed inside a VPC.`,
          },
          {
            id: 'W92',
            reason: `This function supports infrastructure deployment and does not require setting ReservedConcurrentExecutions.`,
          },
        ],
      },
    };
  }
}
