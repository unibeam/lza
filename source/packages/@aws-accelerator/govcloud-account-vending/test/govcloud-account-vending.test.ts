/**
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance
 *  with the License. A copy of the License is located at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions
 *  and limitations under the License.
 */

import * as cdk from 'aws-cdk-lib';
// import { Template } from 'aws-cdk-lib/assertions';
import { GovCloudAccountVendingStack } from '../lib/govcloud-avm-stack';
import { snapShotTest } from './snapshot-test';
// Test prefix
const testNamePrefix = 'Stack(GovCloudAccountVendingStack): ';
const stack = new GovCloudAccountVendingStack(new cdk.App(), 'AWSAccelerator-Test-GovCloudAccountVendingStack', {
  synthesizer: new cdk.DefaultStackSynthesizer({
    generateBootstrapVersionRule: false,
  }),
  acceleratorPrefix: 'AWSAccelerator',
});
/**
 * GovCloudAccountVendingStack construct test
 */
describe('GovCloudAccountVendingStack', () => {
  snapShotTest(testNamePrefix, stack);
});
