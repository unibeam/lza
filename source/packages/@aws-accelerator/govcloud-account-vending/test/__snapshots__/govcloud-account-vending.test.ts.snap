// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GovCloudAccountVendingStack Stack(GovCloudAccountVendingStack):  Snapshot Test 1`] = `
{
  "Resources": {
    "GovCloudAccountVendingFunctionB10FB00E": {
      "DependsOn": [
        "GovCloudAccountVendingFunctionServiceRoleDefaultPolicy02E0FF6C",
        "GovCloudAccountVendingFunctionServiceRole3F1EF782",
      ],
      "Metadata": {
        "cfn_nag": {
          "rules_to_suppress": [
            {
              "id": "W58",
              "reason": "CloudWatch Logs are enabled in AWSLambdaBasicExecutionRole",
            },
            {
              "id": "W89",
              "reason": "This function supports infrastructure deployment and is not deployed inside a VPC.",
            },
            {
              "id": "W92",
              "reason": "This function supports infrastructure deployment and does not require setting ReservedConcurrentExecutions.",
            },
          ],
        },
      },
      "Properties": {
        "Code": {
          "ZipFile": "/**
 *  Copyright 2022 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance
 *  with the License. A copy of the License is located at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions
 *  and limitations under the License.
 */

const {
  OrganizationsClient,
  CreateGovCloudAccountCommand,
  DescribeCreateAccountStatusCommand,
} = require('@aws-sdk/client-organizations');
const { ConfiguredRetryStrategy } = require('@aws-sdk/util-retry');
const cfn = require('cfn-response');

const org = new OrganizationsClient({
  retryStrategy: new ConfiguredRetryStrategy(10, attempt => 100 + attempt * 1000),
  region: 'us-east-1',
});
/**
 * create-govcloud-account - lambda handler
 *
 * @param event
 * @returns cfn-response
 */

exports.handler = async (event, context) => {
  console.log('Received event:\\n' + JSON.stringify(event, null, 2));
  try {
    var acc = event.ResourceProperties.accountName;
    var em = event.ResourceProperties.emailAddress;
    var role = event.ResourceProperties.orgAccessRole;
    let i = 0;

    if (event.RequestType === 'Create') {
      var accResp = await org.send(new CreateGovCloudAccountCommand({ AccountName: acc, Email: em, RoleName: role }));
      console.log(JSON.stringify(accResp));
      var car = accResp.CreateAccountStatus.Id;
      let accStatR = await org.send(new DescribeCreateAccountStatusCommand({ CreateAccountRequestId: car }));
      let accStat = accStatR.CreateAccountStatus.State;
      while (accStat === 'IN_PROGRESS' && i < 40) {
        await new Promise(resolve => setTimeout(resolve, 15e3));
        accStatR = await org.send(new DescribeCreateAccountStatusCommand({ CreateAccountRequestId: car }));
        accStat = accStatR.CreateAccountStatus.State;
        i++;
        // print responses to help troubleshoot
        console.log(\`Attempt: \${i} of 40\`);
        console.log(JSON.stringify(accStatR));
        console.log(accStat);
      }
      if (i === 40) {
        console.log('Timed out');
        return await cfn.send(event, context, 'FAILED');
      } else if (accStat === 'FAILED') {
        var physicalResourceId = accStatR.CreateAccountStatus.FailureReason;
        return await cfn.send(event, context, 'FAILED', physicalResourceId);
      } else if (accStat === 'SUCCEEDED') {
        var responseData = {
          AccountId: accStatR.CreateAccountStatus.AccountId,
          GovCloudAccountId: accStatR.CreateAccountStatus.GovCloudAccountId,
        };
        return await cfn.send(event, context, 'SUCCESS', responseData);
      }
    } else if (event.RequestType === 'Delete' || event.RequestType === 'Update') {
      return await cfn.send(event, context, 'SUCCESS');
    }
  } catch (err) {
    let errMsg = \`\${err.name}:\\n\${err.message}\`;
    let responseData = { Error: errMsg };
    console.log(errMsg);

    return await cfn.send(event, context, 'FAILED', responseData);
  }
};
",
        },
        "Description": "Create AWS GovCloud (US) Accounts",
        "FunctionName": "AWSAccelerator-GovCloudAccountVending",
        "Handler": "index.handler",
        "Role": {
          "Fn::GetAtt": [
            "GovCloudAccountVendingFunctionServiceRole3F1EF782",
            "Arn",
          ],
        },
        "Runtime": "nodejs18.x",
        "Timeout": 900,
      },
      "Type": "AWS::Lambda::Function",
    },
    "GovCloudAccountVendingFunctionServiceRole3F1EF782": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "GovCloudAccountVendingFunctionServiceRoleDefaultPolicy02E0FF6C": {
      "Metadata": {
        "cfn_nag": {
          "rules_to_suppress": [
            {
              "id": "W12",
              "reason": "IAM policy should not allow * resource.",
            },
          ],
        },
      },
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "organizations:CreateGovCloudAccount",
                "organizations:DescribeCreateAccountStatus",
              ],
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "GovCloudAccountVendingFunctionServiceRoleDefaultPolicy02E0FF6C",
        "Roles": [
          {
            "Ref": "GovCloudAccountVendingFunctionServiceRole3F1EF782",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "GovCloudAccountVendingPortfolioEB3BE033": {
      "Properties": {
        "DisplayName": "Landing Zone Accelerator on AWS",
        "ProviderName": "AWS Solutions",
      },
      "Type": "AWS::ServiceCatalog::Portfolio",
    },
    "GovCloudAccountVendingPortfolioPortfolioProductAssociationf14ebeee1f152D65EDA3": {
      "Properties": {
        "PortfolioId": {
          "Ref": "GovCloudAccountVendingPortfolioEB3BE033",
        },
        "ProductId": {
          "Ref": "GovCloudAccountVendingProduct2B8769D5",
        },
      },
      "Type": "AWS::ServiceCatalog::PortfolioProductAssociation",
    },
    "GovCloudAccountVendingProduct2B8769D5": {
      "Properties": {
        "Name": "Landing Zone Accelerator on AWS - AWS GovCloud (US) Account Vending",
        "Owner": "AWS Solutions",
        "ProvisioningArtifactParameters": [
          {
            "Description": "AWS GovCloud (US) Account Vending Product. Create AWS GovCloud (US) accounts. Required inputs are Account name, email and Organization Access Role.",
            "DisableTemplateValidation": false,
            "Info": {
              "LoadTemplateFromURL": {
                "Fn::Sub": "REPLACED-JSON-PATH.json",
              },
            },
            "Name": "v1.0.0",
          },
        ],
      },
      "Type": "AWS::ServiceCatalog::CloudFormationProduct",
    },
  },
}
`;
